File: src/backend/studio-reservation/src/main/java/org/taonaw/studio_reservation/domain/model/staffAccount/StaffAccount.java
package org.taonaw.studio_reservation.domain.model.staffAccount;

import lombok.NonNull;
import org.taonaw.studio_reservation.domain.model.shared.HashedPassword;
import org.taonaw.studio_reservation.domain.model.shared.PasswordEncoder;
import org.taonaw.studio_reservation.domain.model.shared.PersonName;
import org.taonaw.studio_reservation.domain.model.shared.PlainTextPassword;

import java.util.Objects;

public class StaffAccount {
    private StaffAccountId id;
    private PersonName name;
    private LoginId loginId;
    private HashedPassword password;

    public static StaffAccount create(
            @NonNull PersonName staffName,
            @NonNull LoginId loginId,
            @NonNull HashedPassword hashedPassword) {

        var instance = new StaffAccount();
        instance.id = StaffAccountId.newId();
        instance.name = staffName;
        instance.loginId = loginId;
        instance.password = hashedPassword;
        return instance;
    }

    public boolean authenticate(
            @NonNull LoginId loginId,
            @NonNull PlainTextPassword plainTextPassword,
            @NonNull PasswordEncoder passwordEncoder) {

        return this.loginId.equals(loginId) && plainTextPassword.matches(this.password, passwordEncoder);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        StaffAccount that = (StaffAccount) o;
        return id.equals(that.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
}

File: src/backend/identityaccess/src/main/kotlin/org/taonaw/identityaccess/domain/model/staffAccount/StaffAccount.kt
package org.taonaw.identityaccess.domain.model.staffAccount

import org.taonaw.identityaccess.domain.model.shared.HashedPassword
import org.taonaw.identityaccess.domain.model.shared.PasswordHashingService
import org.taonaw.identityaccess.domain.model.shared.PlainTextPassword

class StaffAccount private constructor(
        val staffAccountId: StaffAccountId,
        val staffName: StaffName,
        val userId: UserId,
        val password: HashedPassword) {

    companion object {
        fun create(
                staffName: StaffName,
                userId: UserId,
                plainTextPassword: PlainTextPassword,
                passwordHashingService: PasswordHashingService): StaffAccount {

            return StaffAccount(StaffAccountId.newId(), staffName, userId, plainTextPassword.hash(passwordHashingService))
        }
    }

    fun change(staffName: StaffName, userId: UserId): StaffAccount {
        return StaffAccount(staffAccountId, staffName, userId, password)
    }

    fun changePassword(
            oldPlainTextPassword: PlainTextPassword,
            newPlainTextPassword: PlainTextPassword,
            passwordHashingService: PasswordHashingService): StaffAccount {

        if (!oldPlainTextPassword.matches(password, passwordHashingService))
            StaffAccountOldPasswordDifferentErr().throwErr()

        return StaffAccount(staffAccountId, staffName, userId, newPlainTextPassword.hash(passwordHashingService))
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as StaffAccount

        if (staffAccountId != other.staffAccountId) return false

        return true
    }

    override fun hashCode(): Int {
        return staffAccountId.hashCode()
    }
}